### PR TITLE
Upgrade smart contracts code

### PR DESCRIPTION
**Pull request by andreibancioiu**
- Added `DeductInitialGasForDirectUpgrade()`
- Added extra parameter (code metadata) to `DeployCode()`
- If `upgradeContract` is called, then perform upgrade instead of call

### CODE REVIEW

File: arwen/host/execution.go [hunk 1]
```diff
@@ -274,6 +358,10 @@ func (host *vmHost) execute(input *vmcommon.ContractCallInput) error {
 		return arwen.ErrInitFuncCalledInRun
 	}
 
+	if host.isUpgradeFunctionBeingCalled() {
```
Comment (sasurobert) [inline arwen/host/execution.go]: This should be never reached, why is this protection here?

File: arwen/interface.go [hunk 2]
```diff
@@ -17,13 +17,10 @@ type StateStack interface {
 
 // ArgumentsParser defines the functionality to parse transaction data into arguments and code for smart contracts
 type ArgumentsParser interface {
-	GetArguments() ([][]byte, error)
-	GetCode() ([]byte, error)
+	GetFunctionArguments() ([][]byte, error)
```
Comment (sasurobert) [inline arwen/interface.go]: It is an argument parser. No need to add into the function name. keep it simple

Comment (sasurobert) [inline arwen/interface.go]: Keep it simple, without argument in function name

File: arwen/host/execution.go [hunk 3]
```diff
@@ -144,6 +144,85 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
 	return vmOutput
 }
 
+func (host *vmHost) doRunSmartContractUpgrade(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput) {
```
Comment (sasurobert) [inline arwen/host/execution.go]: This is too similar to the normal deploy func

File: arwen/host/execution.go [hunk 4]
```diff
@@ -144,6 +144,85 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
...
+
+	contractCode, err := input.GetCodeUpgrade()
+	if err != nil {
+		output.SetReturnCode(vmcommon.UpgradeFailed)
+		return vmOutput
+	}
+
+	contractCodeMetadata, err := input.GetCodeMetadataUpgrade()
+	if err != nil {
+		output.SetReturnCode(vmcommon.UpgradeFailed)
```
Comment (camilbancioiu) [inline arwen/host/execution.go]: If you choose to refactor this into a function `performCodeDeploy()`: it will return a `vmOutput` which, in case of failure, contain will contain the return code `vmcommon.ExecutionFailed`. But `doRunSmartContractUpgrade()` could alter the return code from `ExecutionFailed` to `UpgradeFailed`, for this specific case.

File: arwen/interface.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/interface.go]: It makes sense. Now the argument parser distinguishes between function arguments and constructor / deploy arguments. Before this, it still worked, but just because `GetFunction()` returned `parser.code`, which was horrible (by coincidence, the code and the function name were at the same position in the @argument string).

File: arwen/host/execution.go [hunk 5]
```diff
@@ -144,6 +144,85 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
...
+			} else {
+				message = err.Error()
+			}
+			vmOutput = output.CreateVMOutputInCaseOfError(output.ReturnCode(), message)
+		}
+	}()
+
+	address := input.RecipientAddr
+
+	contractCode, err := input.GetCodeUpgrade()
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go]: I suggest moving `GetCodeUpgrade` to `runtimeContext.GetCodeUpgradeFromArgs()`, instead of a method of `ContractCallInput`. It's slightly confusing.

File: arwen/host/execution.go [hunk 6]
```diff
@@ -144,6 +144,85 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
...
+	if err != nil {
+		output.SetReturnCode(vmcommon.UpgradeFailed)
+		return vmOutput
+	}
+
+	contractCodeMetadata, err := input.GetCodeMetadataUpgrade()
```
Comment (camilbancioiu) [inline arwen/host/execution.go]: See comment on `GetCodeUpgrade` above.

File: arwen/interface.go (hunk 2 above)
Comment (andreibancioiu) [reply arwen/interface.go:20]: Yes, now we need to differentiate between constructor and regular function arguments. Name should be explicit. Name should not be ambiguous. Name is simple :)

Comment (andreibancioiu) [reply arwen/interface.go:20]: See above comment.

File: arwen/contexts/metering.go [hunk 7]
```diff
@@ -141,6 +141,20 @@ func (context *meteringContext) DeductInitialGasForIndirectDeployment(input *vmc
 	)
 }
 
+// DeductInitialGasForDirectUpgrade deducts gas for the upgrade of a contract initiated by a Transaction
+func (context *meteringContext) DeductInitialGasForDirectUpgrade(input *vmcommon.ContractCallInput) error {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/contexts/metering.go:145]: Fixed, no extra deduct gas function.

File: arwen/host/execution.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/host/execution.go:171]: Added `GetCodeUpgradeFromArgs()`.

File: arwen/host/execution.go (hunk 6 above)
Comment (andreibancioiu) [reply arwen/host/execution.go:177]: Fixed.

File: arwen/host/execution.go (hunk 3 above)
Comment (andreibancioiu) [reply arwen/host/execution.go]: Extracted common code to another function.

File: arwen/host/execution.go (hunk 4 above)
Comment (andreibancioiu) [reply arwen/host/execution.go]: Refactored, fixed.

File: arwen/contexts/reservedFunctions_test.go [hunk 8]
```diff
@@ -11,9 +12,11 @@ func TestReservedFunctions_IsFunctionReserved(t *testing.T) {
...
 	require.False(t, reserved.IsReserved("foo"))
 	require.True(t, reserved.IsReserved("claimDeveloperRewards"))
+	require.True(t, reserved.IsReserved(arwen.UpgradeFunctionName))
 }
 
 func TestReservedFunctions_GetReserved(t *testing.T) {
 	reserved := NewReservedFunctions()
-	require.ElementsMatch(t, []string{"claimDeveloperRewards"}, reserved.GetReserved())
+	actualReserved := []string{"claimDeveloperRewards", arwen.UpgradeFunctionName}
+	require.ElementsMatch(t, actualReserved, reserved.GetReserved())
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/contexts/reservedFunctions_test.go]: Will be done in a future PR. [EN-6059](https://elrondnetwork.atlassian.net/browse/EN-6059).

File: arwen/contexts/reservedFunctions_test.go [hunk 9]
```diff
@@ -11,9 +12,11 @@ func TestReservedFunctions_IsFunctionReserved(t *testing.T) {
 
 	require.False(t, reserved.IsReserved("foo"))
 	require.True(t, reserved.IsReserved("claimDeveloperRewards"))
+	require.True(t, reserved.IsReserved(arwen.UpgradeFunctionName))
 }
 
 func TestReservedFunctions_GetReserved(t *testing.T) {
 	reserved := NewReservedFunctions()
-	require.ElementsMatch(t, []string{"claimDeveloperRewards"}, reserved.GetReserved())
+	actualReserved := []string{"claimDeveloperRewards", arwen.UpgradeFunctionName}
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/contexts/reservedFunctions_test.go]: Will be done in a future PR. [EN-6059](https://elrondnetwork.atlassian.net/browse/EN-6059).

File: arwen/host/execution.go (hunk 1 above)
Comment (andreibancioiu) [reply arwen/host/execution.go]: Removed constraint.

In order to avoid conflicts with work in progress regarding indirect execution, SC upgrade initiated by parent SC will be done in a future PR. [EN-6058](
https://elrondnetwork.atlassian.net/browse/EN-6058).

File: arwen/contexts/reservedFunctions.go [hunk 10]
```diff
@@ -13,6 +15,7 @@ func NewReservedFunctions(scAPINames []string) *ReservedFunctions {
 
 	var empty struct{}
 	result.functionNames["claimDeveloperRewards"] = empty
```
Comment (sasurobert) [inline arwen/contexts/reservedFunctions.go:17]: ohhh. Needs the map from Elrond-Go but probably in another PR.

Comment (andreibancioiu) [reply arwen/contexts/reservedFunctions.go:17]: Yes, will do 2 PRs (Arwen plus Node).

File: arwen/host/execution.go [hunk 11]
```diff
@@ -11,70 +11,129 @@ func (host *vmHost) doRunSmartContractCreate(input *vmcommon.ContractCreateInput
...
+	runtime := host.Runtime()
+	output := host.Output()
+	storage := host.Storage()
+
+	var err error
+	defer func() {
+		vmOutput = host.onExitDirectCreateOrCall(err, vmOutput)
+	}()
+
+	runtime.SetVMInput(&input.VMInput)
```
Comment (camilbancioiu) [inline arwen/host/execution.go:119]: Use `runtime.InitStateFromContractCallInput()` instead, which also initializes some extra fields of the `runtimeContext`. See `doRunSmartContractCall()` for an example of context initialization.

File: arwen/host/execution.go [hunk 12]
```diff
@@ -224,42 +274,41 @@ func (host *vmHost) CreateNewContract(input *vmcommon.ContractCreateInput) ([]by
 	gasForDeployment := runtime.GetVMInput().GasProvided
 	err = runtime.CreateWasmerInstance(input.ContractCode, gasForDeployment)
 	if err != nil {
-		runtime.PopInstance()
-		runtime.PopState()
-		arwen.RemoveHostContext(idContext)
+		host.onExitIndirectCreate(idContext)
```
Comment (camilbancioiu) [inline arwen/host/execution.go:277]: Please revert the changes in `CreateNewContract()`. The exiting tasks will be very specific to success vs. error, and I need them explicit here. There will be some refactoring afterwards though (`host.onExitIndirectCreate()` will likely exist, but in a different form).

File: arwen/host/execution.go [hunk 13]
```diff
@@ -297,20 +346,13 @@ func (host *vmHost) execute(input *vmcommon.ContractCallInput) error {
 
 	runtime.SetInstanceContextID(idContext)
 
-	function, err := runtime.GetFunctionToCall()
+	err = host.callSCMethodIndirect()
```
Comment (camilbancioiu) [inline arwen/host/execution.go:349]: :+1:

File: arwen/contexts/reservedFunctions_test.go [hunk 14]
```diff
@@ -23,5 +25,5 @@ func TestReservedFunctions_IsFunctionReservedExplicit(t *testing.T) {
 
 func TestReservedFunctions_GetReserved(t *testing.T) {
 	reserved := NewReservedFunctions([]string{})
-	require.ElementsMatch(t, []string{"claimDeveloperRewards"}, reserved.GetReserved())
+	require.ElementsMatch(t, []string{"claimDeveloperRewards", arwen.UpgradeFunctionName}, reserved.GetReserved())
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/contexts/reservedFunctions_test.go:28]: Removed.

File: arwen/host/execution.go (hunk 11 above)
Comment (andreibancioiu) [reply arwen/host/execution.go:119]: Fixed.

File: arwen/host/execution.go (hunk 12 above)
Comment (andreibancioiu) [reply arwen/host/execution.go]: Reverted.
